rfcs/0091-pure-object-prototypes.md

```diff
@@ -0,0 +1,418 @@
+---
+feature: pure-object-prototypes
+start-date: 2021-04-08
+author: Francois-Rene Rideau (@fare)
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues: [lib.experimental](https://github.com/NixOS/rfcs/pull/82)
+---
+
+# Summary
+[summary]: #summary
+
+We propose to add Pure Object Prototypes, or POP, an object system, to the nixpkgs library.
+POP improves upon current Nix extension systems
+by supporting multiple inheritance and default values.
+These improvements solve modularity issues with existing extension systems.
+We actually use POP in production in a fork of nixpkgs to support
+local changes to some packages, by multiple people, with dependencies between changes.
+
+POP can be also made to interoperate both ways with existing extension systems,
+with simple adapters.
+
+
+# Motivation
+[motivation]: #motivation
+
+## A maze of twisty little extension systems, all alike
+
+Nixpkgs features too many mutually incompatible variants
+of what is essentially the same extension system.
+Just between `lib.fixedPoints` and `lib.customisation`, there are already
+`fix'`, `makeExtensibleWithCustomName`, `makeExtensible`,
+`makeOverridable`, `makeScope`, `makeScopeWithSplicing`,
+plus add-ons like `overrideDerivation`.
+Then, many languages and subsystems invent their own variant extension system.
+And while I haven't looked at modules in details,
+I'm told they also include a more elaborate variant in the same family.
+
+This maze of mostly similar yet subtly different yet incompatible constructs
+raises the barrier to entry to learning and using Nix.
+Therefore, to end this anarchy... let's create another extension system!
```

Comment:
Maybe it would be more appropriate to say "to end this fragmentation"? [Anarchy](https://en.wikipedia.org/wiki/Anarchy) Seems like a imprecise term for an RFC and inaccurate term in this context.

Reply:
Done!